lightning/src/ln/peer_handler.rs

```diff
@@ -450,26 +463,55 @@ impl<Descriptor: SocketDescriptor, CM: Deref> PeerManager<Descriptor, CM> where
 			let pause_read = match peers.peers.get_mut(peer_descriptor) {
 				None => panic!("Descriptor for read_event is not already known to PeerManager"),
 				Some(peer) => {
-					assert!(peer.pending_read_buffer.len() > 0);
-					assert!(peer.pending_read_buffer.len() > peer.pending_read_buffer_pos);
+					peer.pending_read_buffer.extend_from_slice(&data);
+					while peer.pending_read_buffer.len() > 0 {
 
-					let mut read_pos = 0;
-					while read_pos < data.len() {
-						{
-							let data_to_copy = cmp::min(peer.pending_read_buffer.len() - peer.pending_read_buffer_pos, data.len() - read_pos);
-							peer.pending_read_buffer[peer.pending_read_buffer_pos..peer.pending_read_buffer_pos + data_to_copy].copy_from_slice(&data[read_pos..read_pos + data_to_copy]);
-							read_pos += data_to_copy;
-							peer.pending_read_buffer_pos += data_to_copy;
+						let mut conduit_option = None;
+						let mut remote_pubkey_option = None;
+
+						if let &mut PeerState::Handshake(ref mut handshake) = &mut peer.encryptor {
+							let (response, conduit, remote_pubkey) = handshake.process_act(&peer.pending_read_buffer, None).unwrap();
+							peer.pending_read_buffer.drain(..); // we read it all
+
+							if let Some(key) = remote_pubkey {
+								remote_pubkey_option = Some(key);
+							}
+
+							peer.pending_outbound_buffer.push_back(response);
+							if let Some(conduit) = conduit {
+								conduit_option = Some(conduit);
+							}
 						}
 
-						if peer.pending_read_buffer_pos == peer.pending_read_buffer.len() {
-							peer.pending_read_buffer_pos = 0;
+						if let Some(key) = remote_pubkey_option {
+							peer.their_node_id = Some(key);
+						}
+
+						if let Some(conduit) = conduit_option {
```

Comment:
Can't all this be done within the `PeerState::Handshake` block above? Then you wouldn't need `conduit_option` and `remote_pubkey_option` there with the de-structured moves in the `if let` statements above. You would just place this code there instead.

Reply:
Not in rust 1.22

Reply:
Hmm, yeah for conduit at least. At very least you shouldn't have destruct the `Option` when assigning to `conduit_option`. You can simply unconditionally assign:

```rust
conduit_option = conduit;
```